File: wcpelib-bukkit/src/main/java/top/wcpe/wcpelib/bukkit/inventory/listener/InventoryListener.java
package top.wcpe.wcpelib.bukkit.inventory.listener;


import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.*;
import top.wcpe.wcpelib.bukkit.inventory.InventoryPlus;
import top.wcpe.wcpelib.bukkit.inventory.entity.*;
import top.wcpe.wcpelib.bukkit.inventory.listener.inter.*;

/**
 * {@link InventoryPlus}的监听事件
 *
 * @author WCPE
 * @date 2021年4月8日 下午5:13:03
 */
public class InventoryListener implements Listener {

    @EventHandler
    public void inventoryClickEvent(InventoryClickEvent e) {
        if (!(e.getInventory().getHolder() instanceof WcpeLibInventoryHolder))
            return;
        InventoryPlus inventoryPlus = ((WcpeLibInventoryHolder) e.getInventory().getHolder()).getInventoryPlus();
        if (inventoryPlus == null) {
            return;
        }
        InventoryClickEventFunctional invOnClick = inventoryPlus.getOnClick();
        if (invOnClick != null) {
            invOnClick.run(new InventoryClickEventDTO(e, inventoryPlus));
        }
        if (inventoryPlus.isDisDoubleClick() && e.getClick() == ClickType.DOUBLE_CLICK) {
            e.setCancelled(true);
        }
        int rawSlot = e.getRawSlot();
        if (inventoryPlus.getIsLockSlot().contains(rawSlot)) {
            e.setCancelled(true);
            return;
        }
        if (inventoryPlus.isDisClickPlayerGui()) {
            int min = inventoryPlus.getRow() * 9;
            int max = inventoryPlus.getRow() * 9 + 35;
            if (rawSlot >= min && rawSlot <= max) {
                e.setCancelled(true);
            }
        }
        Slot<?> slot = inventoryPlus.getSlotMap().get(rawSlot);
        if (slot == null) {
            return;
        }

        SlotEventFunctional onClick = slot.getOnClick();
        if (onClick == null) {
            if (inventoryPlus.isDisClickNullSlot())
                e.setCancelled(true);
            return;
        }

        onClick.run(new SlotDTO(e, inventoryPlus, slot));
    }

    @EventHandler
    public void inventoryDragEvent(InventoryDragEvent e) {
        if (!(e.getInventory().getHolder() instanceof WcpeLibInventoryHolder))
            return;
        InventoryPlus inventoryPlus = ((WcpeLibInventoryHolder) e.getInventory().getHolder()).getInventoryPlus();
        if (inventoryPlus == null) {
            return;
        }
        InventoryDragEventFunctional onDrag = inventoryPlus.getOnDrag();
        if (onDrag != null) {
            onDrag.run(new InventoryDragEventDTO(e, inventoryPlus));
        }
        if (inventoryPlus.isDisDrag()) {
            e.setCancelled(true);
        }

    }

    @EventHandler
    public void inventoryOpenEvent(InventoryOpenEvent e) {
        if (!(e.getInventory().getHolder() instanceof WcpeLibInventoryHolder))
            return;
        InventoryPlus inventoryPlus = ((WcpeLibInventoryHolder) e.getInventory().getHolder()).getInventoryPlus();
        if (inventoryPlus == null) {
            return;
        }
        inventoryPlus.setClose(false);
        InventoryOpenEventFunctional onOpen = inventoryPlus.getOnOpen();
        if (onOpen == null) {
            return;
        }
        onOpen.run(new InventoryOpenEventDTO(e, inventoryPlus));

    }

    @EventHandler
    public void inventoryCloseEvent(InventoryCloseEvent e) {
        if (!(e.getInventory().getHolder() instanceof WcpeLibInventoryHolder))
            return;
        InventoryPlus inventoryPlus = ((WcpeLibInventoryHolder) e.getInventory().getHolder()).getInventoryPlus();
        if (inventoryPlus == null) {
            return;
        }
        if (inventoryPlus.isClose())
            return;
        inventoryPlus.setClose(true);
        InventoryCloseEventFunctional onClose = inventoryPlus.getOnClose();
        if (onClose == null) {
            return;
        }
        onClose.run(new InventoryCloseEventDTO(e, inventoryPlus));
    }
}

File: nukkit/src/main/kotlin/top/wcpe/wcpelib/nukkit/inventory/InventoryListener.kt
package top.wcpe.wcpelib.nukkit.inventory

import cn.nukkit.Player
import cn.nukkit.event.EventHandler
import cn.nukkit.event.Listener
import cn.nukkit.event.inventory.InventoryClickEvent
import cn.nukkit.event.inventory.InventoryCloseEvent
import cn.nukkit.event.inventory.InventoryOpenEvent
import cn.nukkit.event.inventory.InventoryTransactionEvent
import cn.nukkit.inventory.transaction.action.SlotChangeAction
import top.wcpe.wcpelib.nukkit.inventory.entity.InventoryClickEventDTO
import top.wcpe.wcpelib.nukkit.inventory.entity.InventoryCloseEventDTO
import top.wcpe.wcpelib.nukkit.inventory.entity.InventoryOpenEventDTO

class InventoryListener : Listener {
    @EventHandler
    fun inventoryTransactionEvent(e: InventoryTransactionEvent) {
        var transaction = e.transaction
        var from: SlotChangeAction? = null
        var to: SlotChangeAction? = null
        var who: Player? = null

        for (action in transaction.actions) {
            if (action !is SlotChangeAction) {
                continue
            }
            if (action.inventory.holder is Player) {
                who = action.inventory.holder as Player
            }
            if (from == null) {
                from = action
            } else {
                to = action
            }
        }
        var target: SlotChangeAction? =
            (if (from?.inventory is RawInventory) from else if (to?.inventory is RawInventory) to else null) ?: return
        var inventory = target!!.inventory
        var slot = target!!.slot
        if (inventory.holder !is WcpeLibInventoryHolder) return
        val inventoryPlus: InventoryPlus = (inventory.holder as WcpeLibInventoryHolder).inventoryPlus
        var inventoryClickEventDTO = InventoryClickEventDTO(
            InventoryClickEvent(
                who,
                inventory,
                target.slot,
                target.sourceItem,
                target.targetItem
            ), inventoryPlus
        )
        inventoryPlus.onClick?.let {
            it.accept(inventoryClickEventDTO)
            e.isCancelled = inventoryClickEventDTO.inventoryClickEvent.isCancelled
        }
        if (inventoryPlus.isLockSlot.contains(slot)) {
            e.isCancelled = true
            return
        }

        var slotPlus = inventoryPlus.slotMap[slot]

        slotPlus?.listener?.accept(inventoryClickEventDTO)
        e.isCancelled =inventoryClickEventDTO.inventoryClickEvent.isCancelled
    }


    @EventHandler
    fun inventoryOpenEvent(e: InventoryOpenEvent) {
        if (e.inventory.holder !is WcpeLibInventoryHolder) return
        val inventoryPlus: InventoryPlus = (e.inventory.holder as WcpeLibInventoryHolder).inventoryPlus

        inventoryPlus.close = false
        inventoryPlus.onOpen?.let { it.accept(InventoryOpenEventDTO(e, inventoryPlus)) }
    }

    @EventHandler
    fun inventoryCloseEvent(e: InventoryCloseEvent) {
        if (e.inventory.holder !is WcpeLibInventoryHolder) return
        val inventoryPlus: InventoryPlus = (e.inventory.holder as WcpeLibInventoryHolder).inventoryPlus

        if (inventoryPlus.close) return
        inventoryPlus.close = true
        inventoryPlus.onClose?.let { it.accept(InventoryCloseEventDTO(e, inventoryPlus)) }
    }
}